app/src/main/java/openfoodfacts/github/scrachx/openfood/views/splash/SplashActivity.java

```diff
@@ -85,4 +133,237 @@ public void hideLoading(boolean isError) {
+
+        private saveCSVToDb(File file, SharedPreferences settings, Context context) {
+            this.file = file;
+            this.settings = settings;
+            this.context = context;
+        }
+
+        @Override
+        protected void onPreExecute() {
+            super.onPreExecute();
+        }
+
+        @Override
+        protected void onPostExecute(Boolean is) {
+            super.onPostExecute(is);
+            if (is) {
+                SharedPreferences.Editor editor = settings.edit();
+                editor.putBoolean("is_offline_data_available", true);
+                editor.apply();
+                textLoading.setText(getString(R.string.txtLoading));
+                presenter.refreshData();
+            } else {
+                new MaterialDialog.Builder(SplashActivity.this)
+                        .content(getString(R.string.txtConnectionFailed))
+                        .positiveText(getString(R.string.retry))
+                        .negativeText(R.string.txtPictureNeededDialogNo)
+                        .onPositive((dialog, which) -> doDownload())
+                        .onNegative(new MaterialDialog.SingleButtonCallback() {
+                            @Override
+                            public void onClick(@NonNull MaterialDialog dialog, @NonNull DialogAction which) {
+                                textLoading.setText(getString(R.string.txtLoading));
+                                presenter.refreshData();
+                            }
+                        })
+                        .show();
+            }
+        }
+
+        @Override
+        protected Boolean doInBackground(Void... voids) {
+            OfflineProductDao mOfflineProductDao = Utils.getAppDaoSession(context).getOfflineProductDao();
+            //parsing csv file
+            try (CSVReader reader = new CSVReader(new FileReader(file.getAbsolutePath()), '\t')) {
+                List<String[]> records = reader.readAll();
+                List<OfflineProduct> list = new ArrayList<>();
+                Iterator<String[]> iterator = records.iterator();
+                // To skip header in the csv file
+                iterator.hasNext();
+                long size = records.size();
+                long count = 0;
+                while (iterator.hasNext()) {
+                    String[] record = iterator.next();
+                    OfflineProduct offlineProduct = new OfflineProduct(record[1], record[3], record[0], record[2], record[4]);
+                    list.add(offlineProduct);
+                    if (count % 15000 == 0) {
+                        // saving to db
+                        mOfflineProductDao.insertOrReplaceInTx(list);
+                        list.clear();
+                    }
+                    count++;
+                    Log.d(TAG, "completed " + count + " / " + size);
+                    long finalCount = count;
+                    runOnUiThread(new Runnable() {
```

Comment:
Please override and use onProgressUpdate (Progress... values) callback in order to update the UI during saving, instead of runOnUiThread. You can publishProgress(Progress...) from doInBackground to notify the progress.